## CIRIquant ##

[![Build Status](https://travis-ci.com/Kevinzjy/CIRIquant.svg?branch=master)](https://travis-ci.com/Kevinzjy/CIRIquant)
![GitHub release (latest by date)](https://img.shields.io/github/v/release/Kevinzjy/CIRIquant)
[![The MIT License](https://img.shields.io/badge/license-MIT-orange.svg)](https://github.com/Kevinzjy/CIRIquant/blob/master/LICENSE)
![GitHub All Releases](https://img.shields.io/github/downloads/Kevinzjy/CIRIquant/total)
![SourceForge](https://img.shields.io/sourceforge/dm/ciri/CIRIquant)
![Readthedoc](https://readthedocs.org/projects/ciriquant-cookbook/badge/?version=latest)

CIRIquant is a comprehensive analysis pipeline for circRNA detection and quantification in RNA-Seq data

### Documentation ### 

Documentation is available online at [https://ciriquant-cookbook.readthedocs.io/en/latest/index.html](https://ciriquant-cookbook.readthedocs.io/en/latest/index.html)

### Author ###

Authors: Jinyang Zhang(zhangjinyang@biols.ac.cn), Fangqing Zhao(zhfq@biols.ac.cn)

Maintainer: Jinyang Zhang

### Release Notes ###

- Version 1.1: Added support for stranded library and GFF3 format input.
- Version 1.0: The first released version of CIRIquant.

### License ###

The code is released under the MIT License. See the `LICENSE` file for more detail.

### Citing CIRIquant

- Zhang, J., Chen, S., Yang, J. et al. Accurate quantification of circular RNAs identifies extensive circular isoform switching events. Nat Commun 11, 90 (2020) [doi:10.1038/s41467-019-13840-9](https://doi.org/10.1038/s41467-019-13840-9)
